runtime/src/accounts.rs

```diff
@@ -320,36 +332,79 @@ impl Accounts {
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, account);
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &Arc<RwLock<HashMap<Pubkey, CreditOnlyLock>>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            let credit_only_locks = credit_only_locks.read().unwrap();
+            if locks.contains(k)
+                || (credit_only_locks.contains_key(k)
+                    && *credit_only_locks
+                        .get(&k)
+                        .unwrap()
+                        .lock_count
+                        .lock()
+                        .unwrap()
+                        > 0)
+            {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            }
+        }
+        for k in credit_only_keys.iter() {
             if locks.contains(k) {
                 error_counters.account_in_use += 1;
                 debug!("Account in use: {:?}", k);
                 return Err(TransactionError::AccountInUse);
             }
         }
-        for k in keys {
-            locks.insert(**k);
+
+        for k in credit_debit_keys {
+            locks.insert(*k);
+        }
+        for k in credit_only_keys {
+            let mut credit_only_locks = credit_only_locks.write().unwrap();
+            credit_only_locks
+                .entry(*k)
+                .and_modify(|lock| *lock.lock_count.lock().unwrap() += 1)
+                .or_insert_with(|| CreditOnlyLock {
+                    credits: AtomicU64::new(0),
+                    lock_count: Mutex::new(1),
+                });
         }
+
         Ok(())
     }
 
-    fn unlock_account(tx: &Transaction, result: &Result<()>, locks: &mut HashSet<Pubkey>) {
+    fn unlock_account(
+        tx: &Transaction,
+        result: &Result<()>,
+        locks: &mut HashSet<Pubkey>,
+        credit_only_locks: &Arc<RwLock<HashMap<Pubkey, CreditOnlyLock>>>,
+    ) {
+        let (credit_debit_keys, credit_only_keys) = &tx.message().get_account_keys_by_lock_type();
         match result {
             Err(TransactionError::AccountInUse) => (),
             _ => {
-                for k in &tx.message().account_keys {
+                for k in credit_debit_keys {
                     locks.remove(k);
                 }
+                for k in credit_only_keys {
+                    let mut locks = credit_only_locks.write().unwrap();
```

Comment:
This can be `credit_only_locks.read().unwrap()`, the entry has to exist by this point or there's something wrong

Reply:
Actually in this case, replay_stage freeze(), commit_credits() could have cleared the credit_only_locks hashmap by the time this is called. So here I think we should:

1) Get the read lock by `credit_only_locks.read().unwrap()`
2) Check if the `lock_count` exists, if so decrement it, if not, we just return